### PR TITLE
Fix: hole destruction not being recognized caused by subtractive mode…

### LINKED ISSUE #192: What is the best way to create reinforced concrete geometry?
Using the built-in `concrete_rectangular_section()` method, the geometry is generated using addition set operations. This can result in the following issue, where the `control_point` of the concrete coincides with the `control_point` of a steel bar:

```python
from sectionproperties.pre.pre import Material
import sectionproperties.pre.library.concrete_sections as conc_sec
from sectionproperties.analysis.section import Section

concrete = Material(
    name='Concrete', elastic_modulus=30.1e3, poissons_ratio=0.2, density=2.4e-6,
        yield_strength=32, color='lightgrey'
)
steel = Material(
    name='Steel', elastic_modulus=200e3, poissons_ratio=0.3, density=7.85e-6,
        yield_strength=500, color='grey'
)

geom = conc_sec.concrete_rectangular_section(
    b=1000,
    d=112,
    dia=12,
    n_bar=5,
    n_circle=16,
    cover=50,
    conc_mat=concrete,
    steel_mat=steel,
)

geom.plot_geometry()
geom.create_mesh(5)

sec = Section(geom)
sec.plot_mesh()
```
Figure_1
Figure_2
Figure_3

While you get lucky with the material assignment, the mesh size of 5 is clearly not applied to the concrete region.

I thought doing a subtraction of the reinforcement from the concrete section then adding the reinforcement might work, but you are left with holes at the location of the reinforcement:

```python
from sectionproperties.pre.pre import Material
import sectionproperties.pre.library.concrete_sections as conc_sec
import sectionproperties.pre.library.primitive_sections as ps
from sectionproperties.analysis.section import Section

concrete = Material(
    name='Concrete', elastic_modulus=30.1e3, poissons_ratio=0.2, density=2.4e-6,
        yield_strength=32, color='lightgrey'
)
steel = Material(
    name='Steel', elastic_modulus=200e3, poissons_ratio=0.3, density=7.85e-6,
        yield_strength=500, color='grey'
)

b = 1000
d = 112

conc_geom = ps.rectangular_section(b=b, d=d, material=concrete)
reinf_geom = None

cover = 50
dia = 12
n_bar = 5
n_circle = 16

x_i = cover + dia / 2
spacing = (b - 2 * cover - dia) / (n_bar - 1)

for i in range(n_bar):
    bar = ps.circular_section(d=dia, n=n_circle, material=steel)
    bar = bar.shift_section(x_i + spacing * i, cover + dia / 2)

    conc_geom -= bar

    if i == 0:
        reinf_geom = bar
    else:
        reinf_geom += bar

conc_geom.plot_geometry()
reinf_geom.plot_geometry()
geom = conc_geom + reinf_geom
geom.plot_geometry()
```

Figure_11
Figure_12

Not sure if this is the intended behaviour or if you think there is a better way of constructing this geometry? Obviously a bit of a edge case, but would be great to be able to handle this!

Keen to hear anyone's thoughts on this!

### PR DESCRIPTION
**Pull request by connorferster**
Fixes #192, #193

### CODE REVIEW

**Comment by connorferster** [comment]
Test coverage increased :)

**Comment by robbievanleeuwen** [comment]
> Fixes #192, #193

Does this also fix #194?

**Comment by robbievanleeuwen** [comment]
Hi @connorferster,

I still run into the issue in #194, see below code example (a redefinition of the concrete library using subtraction):

```python
from typing import Union, List
from sectionproperties.pre.library.concrete_sections import primitive_sections
from sectionproperties.pre.pre import Material
from sectionproperties.analysis.section import Section
import sectionproperties.pre.pre as pre
import sectionproperties.pre.geometry as geometry

def concrete_rectangular_section(
    b: float,
    d: float,
    dia: float,
    n_bar: int,
    n_circle: int,
    cover: float,
    area: float = None,
    conc_mat: pre.Material = pre.DEFAULT_MATERIAL,
    steel_mat: pre.Material = pre.DEFAULT_MATERIAL,
) -> geometry.CompoundGeometry:

    if n_bar < 2:
        raise ValueError("Please provide 2 or more steel reinforcing bars.")

    geom = primitive_sections.rectangular_section(b=b, d=d, material=conc_mat)

    x_i = cover + dia / 2
    spacing = (b - 2 * cover - dia) / (n_bar - 1)

    for i in range(n_bar):
        if area:
            bar = primitive_sections.circular_section_by_area(
                area=area, n=n_circle, material=steel_mat
            )
        else:
            bar = primitive_sections.circular_section(
                d=dia, n=n_circle, material=steel_mat
            )

        bar = bar.shift_section(x_offset=x_i + spacing * i, y_offset=cover + dia / 2)

        # make hole for bar
        geom = geom - bar

        # add bar
        geom = geom + bar

    return geom


concrete = Material(
    name='Concrete', elastic_modulus=30.1e3, poissons_ratio=0.2, yield_strength=32,
    density=2.4e-6, color='lightgrey'
)
steel = Material(
    name='Steel', elastic_modulus=200e3, poissons_ratio=0.3, yield_strength=500,
    density=7.85e-6, color='grey'
)

geometry = concrete_rectangular_section(
    b=300, d=600, dia=20, n_bar=3, n_circle=24, cover=30, conc_mat=concrete, steel_mat=steel
)
geometry.create_mesh(mesh_sizes=[500])
geometry.plot_geometry()

sec = Section(geometry)
sec.plot_mesh()
```

The issue is that the `__sub__()` method throws the following exception:
```shell
Traceback (most recent call last):
  File "/Users/robbievanleeuwen/Documents/Programming/section-properties/sectionproperties/pre/geometry.py", line 1053, in __sub__
    if self.assigned_control_point and new_polygon.contains(
AttributeError: 'CompoundGeometry' object has no attribute 'assigned_control_point'
```

File: sectionproperties/pre/geometry.py [hunk 1]
```diff
@@ -1677,7 +1677,11 @@ def compile_geometry(self):
                     )
 
         elif isinstance(unionized_poly, Polygon):
-            inadvertent_holes += Geometry(unionized_poly).holes
+            if Geometry(unionized_poly).holes:
+                inadvertent_holes += Geometry(unionized_poly).holes
```
Comment (robbievanleeuwen) [inline sectionproperties/pre/geometry.py:1681]: This seems to work well :+1:

File: sectionproperties/pre/geometry.py [hunk 2]
```diff
@@ -1872,3 +1876,14 @@ def round_polygon_vertices(poly: Polygon, tol: int) -> Polygon:
...
+
+def check_geometry_overlaps(lop: List[Polygon]) -> bool:
+    """
+    Returns True if any of the Polygon in the list of Polygons, 'lop',
+    are overlapping with any other Polygons in 'lop'. Returns False
+    if all Polygon are disjoint.
+    """
+    union_area = unary_union(lop).area
+    sum_polygons = sum([poly.area for poly in lop])
+    return union_area != sum_polygons
```
Comment (robbievanleeuwen) [inline sectionproperties/pre/geometry.py:1889]: Nice!

**Review by robbievanleeuwen — APPROVED** [review]
Looks good!

**Pull request merged**